### ISSUE TITLE
Support rendering under Jest Storyshots

### ISSUE THREAD

**Issue by bradleyayers** [maslianok/react-resize-detector#34]
I get the following error:

```
    TypeError: Cannot read property 'parentElement' of null

      at ResizeDetector.componentDidMount (node_modules/react-resize-detector/lib/components/ResizeDetector.js:97:105)
      at commitLifeCycles (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:6427:22)
      at commitAllLifeCycles (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7516:7)
      at HTMLUnknownElement.callCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1881:14)
      at invokeEventListeners (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:193:27)
      at HTMLUnknownElementImpl._dispatch (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
      at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
      at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
      at HTMLUnknownElement.dispatchEvent (node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
      at Object.invokeGuardedCallbackDev (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1919:16)
      at invokeGuardedCallback (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:1968:29)
      at commitRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7655:7)
      at completeRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8624:34)
      at performWorkOnRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8568:9)
      at performWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8499:7)
      at performSyncWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8471:3)
      at requestWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8371:5)
      at scheduleWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8248:11)
      at scheduleRootUpdate (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8713:3)
      at updateContainerAtExpirationTime (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8740:10)
      at updateContainer (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8751:10)
      at create (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:9221:5)
      at getRenderedTree (node_modules/@storybook/addon-storyshots/dist/react/renderTree.js:24:14)
      at renderOnly (node_modules/@storybook/addon-storyshots/dist/test-bodies.js:73:16)
      at Object.<anonymous> (node_modules/@storybook/addon-storyshots/dist/index.js:121:20)
```

I think this could be avoided by moving code out of `componentDidMount` into the `div`'s `ref` callback. I'll test this and raise a PR if it pans out.

**Comment by maslianok** [comment]
@bradleyayers Thanks for raising this issue and for creating a PR!

In this issue you're saying that in some environments (e.g. Jest Storyshots) the refs aren't accessible in the componentDidMount? Am I right?

Sorry, I'm not familiar with Jest Storyshots, but it should be possible to mock the ResizeDetector component. I don't think we can move the componentDidMount logic inside the ref callback, because ref callback executed each time component gets updated.

**Comment by yannick-milanetto** [comment]
Hello,

FYI, i get the same error when using Preact version of my component with ReactResizeSensor
This seems to be related to:
https://github.com/developit/preact/issues/477
Would it be possible to find a way not to rely on refs in componentDidMount ?
This would solve, i guess, both the Jest and Preact issues.

Thanks for your answer
Cheers

**Comment by lamhieu-vk** [comment]
hi @yannick-milanetto , you can only get the element size when it is created, but there is no other way to get it!
why use `componentDidMount` to make sure element is mounted, then `polyfill` will use window api to get style and calc to size of element. All package to get size of element don't get size before component mounted, and it not support server-side render!

**Comment by yannick-milanetto** [comment]
hi @lamhieu-vk 
Thanks for your answer. I understand that you need to get element size on componentDidMount, the problem is not really that. It's more related to the fact that in preact there's no guarantee that component's refs are up-to-date when entering in componentDidMount as explained in the preact issue i've linked above : 

> Currently Preact does not defer ref callback invocation like it does componentDidMount()

This results in ReactResizeDetector breaking at the line 5 of the component logic :
```
_createClass(ResizeDetector, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var resizableElementId = this.props.resizableElementId;
      var resizableElement = resizableElementId ? document.getElementById(resizableElementId) : this.el.parentElement;
      this.ro.observe(resizableElement);
    }
  }
```
At this point, this.el is undefined, hence this.el.parentElement throws an error : 
`VM2637 WUuE:148 Uncaught (in promise) TypeError: Cannot read property 'parentElement' of undefined
    at ResizeDetector.componentDidMount (VM2394 WUuE:148)`

Unfortunately i can't find a way to make it work with preact :(

**Comment by yannick-milanetto** [comment]
Hi maslianok,
I tried the resizeElementId approach and indeed it does work fine with Preact.
However it breaks the React SSR :
```
Uncaught TypeError: Failed to execute 'observe' on 'ResizeObserver': parameter 1 is not of type 'Element'.
    at ResizeDetector.componentDidMount (index.js?594b:149)
```

With the classic approach (w/o using resizeElementId) React SSR works fine.
I'm still investigating if something can be fixed, either in the way I render my Preact component, or in the way I use your ReactResizeDetector lib.

**Comment by maslianok** [comment]
It makes no sense to render ReactResizeDetector on a server.
I would suggest you add some check to render only in a browser
```
<div id="myId">
  {...}
  {!ssr && <ReactResizeDetector resizableElementId="myId" handleWidth handleHeight onResize={this.onResize} />}
</div>
```

and you can declare the `ssr` variable based on your environment. For example,
```
const ssr = !window;
```

or pass this variable from the webpack.

**Comment by yannick-milanetto** [comment]
Hi maslianok,

Agreed, it totally makes sense. It works fine now.
So I can confirm that for the issue of `this.el` being null / undefined in a Preact context, using resizableElementId provides a working alternative.

For this same issue in a Jest storyshots context, i can't help that much as I don't use Jest storyshots.

Thx

**Comment by lamhieu-vk** [comment]
I fixed at https://github.com/maslianok/react-resize-detector/pull/40

**Comment by maslianok** [comment]
Thanks for fixing this, @lamhieu-vk 

Released in 3.1.1
